### ISSUE TITLE
Calling setInAppMessagesVisible on a non-main thread leads to autolayout engine exceptions 

### ISSUE THREAD

**Issue by michaelhollman** [intercom/intercom-ios#181]
## Version info
  - Intercom for iOS version: 3.0.11
  - iOS version: 10.*

I unfortunately haven't had a chance to test with 3.0.12 yet, due to the fact that our cocoapods are currently a mess with all the swift 2.2 vs 2.3. vs 3.0 migrations and issues.

## Expected behavior

Calling `[Intercom setInAppMessagesVisible:]` for the first time app-wide on a thread other than the main thread should either not lead to running autolayout code on a non-main thread, or proactively warn developers to not do this.

This could be accomplished by some combination of…

- Dispatching view code to the main thread (if/when necessary)
- Short-circuiting if called on any thread but the main thread
- Throwing an assertion failure and/or logging when called on any thread but the main thread
- Explicitly mentioning in the documentation that the method must be called on the main thread

I think is _should_ be obvious to most people with an understanding of UIKit that calling something view-related off the main thread is a bad idea, but given how hard it was to track down the exact cause of the exception below, at least _some_ level of assertion or logging would likely be appreciated by fools unfortunate as myself in the future. 

## Actual behavior

Calling `[Intercom setInAppMessagesVisible:]` for the first time on a non-main thread leads to the following exception:

```obj-c
This application is modifying the autolayout engine from a background thread after the engine was accessed from the main thread. This can lead to engine corruption and weird crashes.
 Stack:(
	0   CoreFoundation                      0x000000010ff3434b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x0000000111b8a21e objc_exception_throw + 48
	2   CoreFoundation                      0x000000010ff9d265 +[NSException raise:format:] + 197
	3   Foundation                          0x000000010c113320 _AssertAutolayoutOnAllowedThreadsOnly + 180
	4   Foundation                          0x000000010c113068 -[NSISEngine _optimizeWithoutRebuilding] + 61
	5   Foundation                          0x000000010bf3ae8c -[NSISEngine optimize] + 91
	6   Foundation                          0x000000010c110f88 -[NSISEngine performPendingChangeNotifications] + 84
	7   UIKit                               0x000000010e2f0cbd -[UIView(CALayerDelegate) _wantsReapplicationOfAutoLayoutWithLayoutDirtyOnEntry:] + 158
	8   UIKit                               0x000000010e2f136a -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 1275
	9   QuartzCore                          0x000000010d91bcdc -[CALayer layoutSublayers] + 146
	10  QuartzCore                          0x000000010d90f7a0 _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 366
	11  QuartzCore                          0x000000010d90f61e _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 24
	12  QuartzCore                          0x000000010d89d62c _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 280
	13  QuartzCore                          0x000000010d8ca713 _ZN2CA11Transaction6commitEv + 475
	14  QuartzCore                          0x000000010d8cab91 _ZN2CA11Transaction14release_threadEPv + 593
	15  libsystem_pthread.dylib             0x00000001135633dc _pthread_tsd_cleanup + 555
	16  libsystem_pthread.dylib             0x0000000113562f63 _pthread_exit + 117
	17  libsystem_pthread.dylib             0x0000000113561582 pthread_attr_getschedpolicy + 0
	18  libsystem_pthread.dylib             0x000000011355f341 start_wqthread + 13
)
```

_Note:_ I have not experimented with subsequent calls to `[Intercom setInAppMessagesVisible:]`. I noticed the use of a `dispatch_once` in the backtraces, so the behavior may be different for subsequent calls. We make a call to `[Intercom setInAppMessagesVisible:NO]` from a dispatched block very early in our application's lifecycle, which was the culprit for us.
 
## Steps to reproduce
 
Shortly after the app launches (e.g. in `application:didFinishLaunchingWithOptions:`), do…

```obj-c
dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
    [Intercom setInAppMessagesVisible:NO];
});
```

## Logs

The exception backtrace in the `Actual behavior` section is what got me started digging into this. I couldn't determine where the `CATransaction` was coming from, so I created a symbolic breakpoint for `CA::Transaction::create()` and monitored every time when it was hit (which was a _lot_) and waited for a case when it was hit off of the main thread. This eventually lead me to a backtrace of…

```obj-c
* thread #7: tid = 0x2f4a9, 0x000000010d8cad04 QuartzCore`CA::Transaction::create(), queue = 'com.apple.root.default-qos', stop reason = breakpoint 2.1
  * frame #0: 0x000000010d8cad04 QuartzCore`CA::Transaction::create()
    frame #1: 0x000000010d8cb357 QuartzCore`CA::Transaction::ensure_compat() + 11
    frame #2: 0x000000010d91b961 QuartzCore`CA::Layer::set_needs_layout() + 29
    frame #3: 0x000000010e2ded7d UIKit`-[UIView(Hierarchy) setNeedsLayout] + 333
    frame #4: 0x000000010e2c52c1 UIKit`-[UIView _invalidateAppearanceForSubviewsOfClass:] + 73
    frame #5: 0x000000010e2c5381 UIKit`-[UIView _invalidateAppearanceForSubviewsOfClass:] + 265
    frame #6: 0x000000010e2c5381 UIKit`-[UIView _invalidateAppearanceForSubviewsOfClass:] + 265
    frame #7: 0x000000010e9f1bf2 UIKit`-[_UIAppearance _handleSetterInvocation:] + 1125
    frame #8: 0x000000010feb9a2e CoreFoundation`___forwarding___ + 526
    frame #9: 0x000000010feb9798 CoreFoundation`__forwarding_prep_0___ + 120
    frame #10: 0x0000000107c88123 RedactedAppName`+[ICMAppearanceReset resetAppearanceWithin:] + 81
    frame #11: 0x0000000107cdc783 RedactedAppName`-[ICMPresentationManager init] + 195
    frame #12: 0x0000000107cdc6ab RedactedAppName`__40+[ICMPresentationManager sharedInstance]_block_invoke + 24
    frame #13: 0x00000001131b80cd libdispatch.dylib`_dispatch_client_callout + 8
    frame #14: 0x000000011319d1fc libdispatch.dylib`dispatch_once_f + 501
    frame #15: 0x0000000107cdc690 RedactedAppName`+[ICMPresentationManager sharedInstance] + 45
    frame #16: 0x0000000107c8f149 RedactedAppName`+[ICMBridge setInAppMessagesVisible:] + 37
    frame #17: 0x0000000106e5e721 RedactedAppName`-[IntercomManager hideNotifications:](self=0x00006080000135a0, _cmd="hideNotifications:", hide=YES) + 337 at IntercomManager.m:209
    frame #18: 0x0000000106da3a41 RedactedAppName`__42-[ThirdPartyServiceManager startIntercom:]_block_invoke((null)=<unavailable>) + 561 at ThirdPartyServiceManager.m:196
    frame #19: 0x000000011318e980 libdispatch.dylib`_dispatch_call_block_and_release + 12
    frame #20: 0x00000001131b80cd libdispatch.dylib`_dispatch_client_callout + 8
    frame #21: 0x0000000113197366 libdispatch.dylib`_dispatch_queue_override_invoke + 1426
    frame #22: 0x00000001131993b7 libdispatch.dylib`_dispatch_root_queue_drain + 720
    frame #23: 0x000000011319908b libdispatch.dylib`_dispatch_worker_thread3 + 123
    frame #24: 0x00000001135614de libsystem_pthread.dylib`_pthread_wqthread + 1129
    frame #25: 0x000000011355f341 libsystem_pthread.dylib`start_wqthread + 13
```

`ThirdPartyServiceManager` and `IntercomManager` are just utility classes that we use to wrap the Intercom SDK (among other things) to make managing user state easier.

The `_dispatch_call_block_and_release` on frame 19 is the block we dispatch with `dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{ … });`.

**Comment by jtreanor** [comment]
Hi @michaelhollman,

Thanks for reporting this and giving so much detail.

I agree that it should be possible to call this (and all) methods from any thread. We will address this in a future release. I'll keep you updated.

For now, of course you can run explicitly on the main thread:
```
dispatch_async(dispatch_get_global_main_queue(), ^{
    [Intercom setInAppMessagesVisible:NO];
});
```

**Comment by michaelhollman** [comment]
@jtreanor Fantastic!

And yeah, that's exactly what we're doing for the time being. Thank you for looking into this!

**Comment by jtreanor** [comment]
This has now been fixed in [3.0.13](https://github.com/intercom/intercom-ios/releases/tag/3.0.13).

Thanks again for reporting it @michaelhollman.
